src/read/pipelining.rs

```diff
@@ -0,0 +1,1209 @@
+//! Pipelined extraction into a filesystem directory.
+
+pub mod path_splitting {
+    use displaydoc::Display;
+    use thiserror::Error;
+
+    use std::collections::BTreeMap;
+
+    use crate::spec::is_dir;
+
+    /// Errors encountered during path splitting.
+    #[derive(Debug, Display, Error)]
+    pub enum PathSplitError {
+        /// entry path format error: {0:?}
+        PathFormat(String),
+        /// file and directory paths overlapped: {0:?}
+        FileDirOverlap(String),
+    }
+
+    fn split_by_separator(entry_path: &str) -> Result<impl Iterator<Item = &str>, PathSplitError> {
+        if entry_path.contains('\\') {
+            if entry_path.contains('/') {
+                return Err(PathSplitError::PathFormat(format!(
+                    "path {:?} contained both '\\' and '/' separators",
+                    entry_path
+                )));
+            }
+            Ok(entry_path.split('\\'))
+        } else {
+            Ok(entry_path.split('/'))
+        }
+    }
+
+    /* TODO: consider using crate::unstable::path_to_string() for this--it involves new
+     * allocations, but that really shouldn't matter for our purposes. I like the idea of using our
+     * own logic here, since parallel/pipelined extraction is really a different use case than the
+     * rest of the zip crate, but it's definitely worth considering. */
+    pub(crate) fn normalize_parent_dirs<'a>(
+        entry_path: &'a str,
+    ) -> Result<(Vec<&'a str>, bool), PathSplitError> {
+        if entry_path.starts_with('/') || entry_path.starts_with('\\') {
+            return Err(PathSplitError::PathFormat(format!(
+                "path {:?} began with '/' or '\\' and is absolute",
+                entry_path
+            )));
+        }
+        let is_dir = is_dir(entry_path);
+
+        let mut ret: Vec<&'a str> = Vec::new();
+        for component in split_by_separator(entry_path)? {
+            match component {
+                /* Skip over repeated separators "//". We check separately for ending '/' with the
+                 * `is_dir` variable. */
+                "" => (),
+                /* Skip over redundant "." separators. */
+                "." => (),
+                /* If ".." is present, pop off the last element or return an error. */
+                ".." => {
+                    if ret.pop().is_none() {
+                        return Err(PathSplitError::PathFormat(format!(
+                        "path {:?} has too many '..' components and would escape the containing dir",
+                        entry_path
+                    )));
+                    }
+                }
+                _ => {
+                    ret.push(component);
+                }
+            }
+        }
+        if ret.is_empty() {
+            return Err(PathSplitError::PathFormat(format!(
+                "path {:?} resolves to the top-level directory",
```

Comment:
This is valid if the destination is the root directory.

Reply:
Ah ok, that makes sense. I'll also need to update this code anyway to address making directories temporarily writable as you mentioned in the other comment.

Reply:
Hmm, actually, clarification: when you say "destination", you're referring to the extraction dir? It's probably not clear, but this method `normalize_parent_dirs()` accepts the entire path string (including final filename component), and uses the `is_dir` boolean tuple argument to indicate whether to split off the final component as a filename. So if we have `ret.is_empty()`, what that means is that the entire path resolves to the equivalent of `./`. Are you saying that `./` directory entries are valid, but we should just ignore them instead of erroring here? That would make sense, just want to make sure.

Reply:
Yes, that's right.